lib/ProgramCall.js

```diff
@@ -31,59 +33,108 @@ class ProgramCall {
+          // eslint-disable-next-line no-continue
+          continue;
+        }
+        // handle nested ds
+        this.xml += '<ds';
+        // append <ds> options
+        if (options.dim) { this.xml += ` dim='${options.dim}'`; }
+        if (options.dou) { this.xml += ` dou='${options.dou}'`; }
+        if (options.len) { this.xml += ` len='${options.len}'`; }
+        if (options.data) { this.xml += ` data='${options.data}'`; }
+        this.xml += '>';
+
+        for (let j = 0; j < ds[i].length; j += 1) {
+          this.__addDsNodes__(ds[i][j]);
+        }
+        this.xml += '</ds>';
       }
-      this.xml += '>';
+    } else {
+      this.__addDataNode__(ds[0], ds[1], ds[2], ds[3]);
     }
+  }
 
-    if (Array.isArray(data)) { // If it is a struct parameter, recursively parse its children.
+  /**
+   * Internal function to handle ds and data within <parm> node
+   * @param {object} parameter
+   */
+  __addData__(parameter = {}) {
+    // adding a data structure with data nodes
+    if (Array.isArray(parameter.value)) {
       this.xml += '<ds';
-      if (opt && typeof opt === 'object') { // append <ds> options
-        if (opt.dim) { this.xml += ` dim='${opt.dim}'`; }
-        if (opt.dou) { this.xml += ` dou='${opt.dou}'`; }
-        if (opt.len) { this.xml += ` len='${opt.len}'`; }
-        if (opt.data) { this.xml += ` data='${opt.data}'`; }
-      }
+      // append <ds> options
+      if (parameter.dim) { this.xml += ` dim='${parameter.dim}'`; }
+      if (parameter.dou) { this.xml += ` dou='${parameter.dou}'`; }
+      if (parameter.len) { this.xml += ` len='${parameter.len}'`; }
+      if (parameter.data) { this.xml += ` data='${parameter.data}'`; }
       this.xml += '>';
-
-      for (let i = 0; i < data.length; i += 1) {
-        this.addParam(data[i][0], data[i][1], data[i][2], true);
-      }
+      this.__addDsNodes__(parameter.value, parameter);
       this.xml += '</ds>';
-    } else { // A simple parameter
-      this.xml += `<data type='${type}'`;
-      if (opt && typeof opt === 'object') { // append <data> options
-        Object.keys(opt).forEach((key) => {
-          this.xml += ` ${key}='${opt[key]}'`;
-        });
-      }
-      this.xml += `>${data}</data>`;
+    } else {
+      this.__addDataNode__(parameter.name, parameter.type, parameter.value, parameter);
     }
+  }
 
-    if (!inDs) { // In recursive mode, if it is an element in DS, then no <parm> or </parm> needed.
-      this.xml += '</parm>';
+  /**
+    * @description adds a parameter to the program XML
+    * @param {object} parmeter
+    */
+  addParam(parameter = {}) {
+    if (typeof parameter !== 'object') {
+      throw new Error('Expected the first parameter to be an object');
     }
+    if (parameter.type === undefined) {
+      throw new Error('Expected \'type\' key on the parameter object');
+    }
+    const nameNode = parameter.name ? ` name='${parameter.name}'` : '';
+    this.xml += `<parm${nameNode}`;
```

Comment:
Please do this like parameter.io and parameter.by are done below.

Reply:
Thanks will do